Unit test support for PHP 8.2

mikkamp opened the pull request:
### Changes proposed in this Pull Request:

This PR prepares support for running our unit tests in PHP 8.2. The main change in this PR is to get rid of the deprecation messages for creating dynamic properties. In our unit tests we've been following that pattern a lot by declaring them as properties within the doc block, but not as actual properties in the class. This PR changes that to move them from the docblock to actual declarations.

In addition to that one of the library packages `google/protobuf` was also generating a substantial amount of warnings. In the latest version they get around that by declaring `#[\AllowDynamicProperties]` in the Message base class. So I've upgraded the library to take advantage of this change:

`composer update google/protobuf --with-all-dependencies`
```
- Upgrading google/protobuf (v3.21.12 => v3.22.2)
```

In short the changes in this PR bring down the errors from 500+, to just 5 errors left when running the unit tests.

Two of the errors are coming from the same error that prevents us from running tests on PHP 8.1, see: https://github.com/woocommerce/google-listings-and-ads/issues/1758#issuecomment-1415897005
It has already been fixed but we are still waiting on a [2.13.2 release](https://github.com/googleapis/google-api-php-client/releases) to take advantage of the fix.

##### Update
2.13.2 has been released so I went ahead and updated the google apiclient package which reduces the error count to just 3.
`composer update google/apiclient --with-all-dependencies`
```
- Upgrading firebase/php-jwt (v6.3.2 => v6.4.0)
- Upgrading google/apiclient (v2.13.0 => v2.13.2)
- Upgrading google/apiclient-services (v0.286.0 => v0.295.0)
- Upgrading google/auth (v1.25.0 => v1.26.0)
- Upgrading guzzlehttp/psr7 (2.4.3 => 2.4.4)
- Upgrading phpseclib/phpseclib (3.0.18 => 3.0.19)
- Upgrading psr/http-client (1.0.1 => 1.0.2)
- Upgrading psr/http-factory (1.0.1 => 1.0.2)
- Upgrading psr/http-message (1.0.1 => 1.1)
```

The other three errors are coming from WooCommerce core, which has already been fixed, but at the moment the [PR is still a draft](https://github.com/woocommerce/woocommerce/pull/37063) and we would have to wait till it's included within a release.

Luckily enough all our code (which is covered by unit tests) did conform to PHP 8.2 standards, so we didn't need any changes there.


### Detailed test instructions:
1. Run `composer install` to upgrade composer packages
2. Install unit tests `bin/install-wp-tests.sh <db> <db_user> <db_pass>` 
3. Run unit tests in PHP 8.2 `/usr/bin/php8.2 vendor/bin/phpunit`
4. Confirm that there are only 5 errors left when running the unit tests with PHP 8.2


### Changelog entry
* Dev - Unit test support for PHP 8.2.

mikkamp committed:
Prevent dynamic properties in test classes

mikkamp committed:
Update google/protobuf package

eason9487 commented on tests/Unit/Ads/AssetSuggestionsServiceTest.php:
```diff
@@ -25,28 +25,59 @@
+	/** @var MockObject|ImageUtility $image_utility */
+	protected $image_utility;
+
+	/** @var MockObject|AdsAssetGroupAsset asset_group_asset */
```
Looks like it missed the leading sign `$`?

eason9487 commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -14,14 +14,21 @@
+	/** @var MockObject|BudgetRecommendationQuery $budget_recommendation_query */
+	protected $budget_recommendation_query;
+
+	/** @var MockObject|ISO3166DataProvider $iso_provider; */
```
There is an extra semicolon at the end.

eason9487 commented on tests/Unit/MerchantCenter/MerchantCenterServiceTest.php:
```diff
@@ -30,28 +30,59 @@
+	/** @var MockObject|TargetAudience $target_audience */
+	protected $target_audience;
+
+	/** @var MockObject|TransientsInterface  $transients */
```
It probably doesn't matter at all but this is the only case that has two spaces.

eason9487 reviewed (approved):
Thanks for the work. LGTM.

Tested locally and got the same result of 5 errors.

![2023-04-12 17 18 24](https://user-images.githubusercontent.com/17420811/231430189-359f54b7-65e9-4759-ae0d-84cf27e30c44.png)

---

Noticed three inconsistent formatting in PHPDoc but they should be very minor things.

> It has already been fixed but we are still waiting on a [1.13.2 release](https://github.com/googleapis/google-api-php-client/releases) to take advantage of the fix.

I guess it was supposed to mean the [2.13.2 release](https://github.com/googleapis/google-api-php-client/releases/tag/v2.13.2)? If yes, it happened to be released last week, and would the upgrade of `google/apiclient` be handled along with this PR as well?

mikkamp committed:
Fix docblock comments for unit tests

mikkamp committed:
Merge develop into dev/unit-tests-php-8-2-support

mikkamp committed:
Update additional redirect test

mikkamp committed:
Update Google apiclient package

mikkamp commented:
Thanks @eason9487 for catching those inconsistencies in the docblock comments, I've gone ahead and updated them.

> I guess it was supposed to mean the [2.13.2 release](https://github.com/googleapis/google-api-php-client/releases/tag/v2.13.2)?

Yes it was 2.13.2, I've gone ahead and updated the `google/apiclient` package:
`composer update google/apiclient --with-all-dependencies`
```
- Upgrading firebase/php-jwt (v6.3.2 => v6.4.0)
- Upgrading google/apiclient (v2.13.0 => v2.13.2)
- Upgrading google/apiclient-services (v0.286.0 => v0.295.0)
- Upgrading google/auth (v1.25.0 => v1.26.0)
- Upgrading guzzlehttp/psr7 (2.4.3 => 2.4.4)
- Upgrading phpseclib/phpseclib (3.0.18 => 3.0.19)
- Upgrading psr/http-client (1.0.1 => 1.0.2)
- Upgrading psr/http-factory (1.0.1 => 1.0.2)
- Upgrading psr/http-message (1.0.1 => 1.1)
```

It's now down to just 3 errors. It's ready for another round of review if you could confirm that you are getting the same results.
I also ran it with PHP 8.1 which now doesn't produce any errors during the tests: `/usr/bin/php8.1 vendor/bin/phpunit`

eason9487 reviewed (approved):
Thanks for the update.

> It's now down to just 3 errors. It's ready for another round of review if you could confirm that you are getting the same results.

Confirmed the new result is the same on my end.

```
...............................................................  63 / 922 (  6%)
............................................................... 126 / 922 ( 13%)
............................................................... 189 / 922 ( 20%)
.........................................................E..... 252 / 922 ( 27%)
............................................................... 315 / 922 ( 34%)
............................................................... 378 / 922 ( 40%)
............................................................... 441 / 922 ( 47%)
............................................................... 504 / 922 ( 54%)
............................................................... 567 / 922 ( 61%)
............................................................... 630 / 922 ( 68%)
............................................................... 693 / 922 ( 75%)
............................................................... 756 / 922 ( 81%)
............................................................... 819 / 922 ( 88%)
.................................................EE............ 882 / 922 ( 95%)
........................................                        922 / 922 (100%)

Time: 00:44.201, Memory: 145.00 MB

There were 3 errors:
```
